Source/WebCore/html/LinkRelAttribute.cpp

```diff
@@ -40,82 +41,79 @@
 
 namespace WebCore {
 
-// Keep LinkRelAttribute::isSupported() in sync when updating this constructor.
+// https://html.spec.whatwg.org/#linkTypes
+
+struct LinkType {
+    ASCIILiteral name;
+    bool (*isEnabled)(const Document&);
+    void (*updateRel)(LinkRelAttribute&);
+};
+
+static constexpr std::array<const LinkType, 12> linkTypes { {
+    { "alternate"_s, [](auto) { return true; }, [](auto relAttribute) { relAttribute.isAlternate = true; } },
+    { "apple-touch-icon"_s, [](auto) { return true; }, [](auto relAttribute) { relAttribute.iconType = LinkIconType::TouchIcon; } },
+    { "apple-touch-icon-precomposed"_s, [](auto) { return true; }, [](auto relAttribute) { relAttribute.iconType = LinkIconType::TouchPrecomposedIcon; } },
+    { "dns-prefetch"_s, [](auto) { return true; }, [](auto relAttribute) { relAttribute.isDNSPrefetch = true; } },
+    { "expect"_s, [](auto) { return true; }, [](auto relAttribute) { relAttribute.isInternalResourceLink = true; } },
+    { "icon"_s, [](auto) { return true; }, [](auto relAttribute) { relAttribute.iconType = LinkIconType::Favicon; } },
+#if ENABLE(APPLICATION_MANIFEST)
+    { "manifest"_s, [](auto) { return true; }, [](auto relAttribute) { relAttribute.isApplicationManifest = true; } },
+#else
+    { "manifest"_s, [](auto) { return false; }, [](auto) { } },
+#endif
+    { "modulepreload"_s, [](auto document) { return document.settings().linkModulePreloadEnabled(); }, [](auto relAttribute) { relAttribute.isLinkModulePreload = true; } },
+    { "preconnect"_s, [](auto document) { return document.settings().linkPreconnectEnabled(); }, [](auto relAttribute) { relAttribute.isLinkPreconnect = true; } },
+    { "prefetch"_s, [](auto document) { return document.settings().linkPrefetchEnabled(); }, [](auto relAttribute) { relAttribute.isLinkPrefetch = true; } },
+    { "preload"_s, [](auto document) { return document.settings().linkPreloadEnabled(); }, [](auto relAttribute) { relAttribute.isLinkPreload = true; } },
+    { "stylesheet"_s, [](auto) { return true; }, [](auto relAttribute) { relAttribute.isStyleSheet = true; } },
+} };
+
+static std::optional<const LinkType> findLinkType(StringView rel)
+{
+    for (auto linkType : linkTypes) {
```

Comment:
I think this is the kind of thing we usually use `SortedArrayMap` (+ `ComparableLettersLiteral` for). cc @darinadler

Reply:
Interesting! That does have this comment at the top indicating it might not be a win given the current size:
```
class SortedArrayBase {
protected:
    // Some informal empirical tests indicate that arrays shorter than this are faster to
    // search with linear search than with binary search. Even if we don't get this threshold
    // exactly right, it's helpful for both performance and code size to use linear search at
    // least for very small arrays, and important for performance to make sure that we use
    // binary search for much larger ones.
    static constexpr size_t binarySearchThreshold = 20;
};
```
But if I'm correct about that I should probably at least add a comment indicating this should be changed to `SortedArrayMap` once we hit that number.

Reply:
I don't think it matters, it is still nicer than doing your own linear search with case insensitive matching I think.

Reply:
That is fair. Certainly good to rely on more fundamental primitives.

Reply:
One of the best features of `SortedArrayMap` is that it automatically starts using the better algorithm as the table gets longer and does a linear search when that's more efficient. So I suggest using it without thinking twice about the size of the table

Reply:
@darinadler : Maybe the comment in SortedArrayBase should be updated accordingly?

Reply:
Sure, I’ll do that.